### ISSUE TITLE
Refactor to support aspnetcore middleware and httpmodule versions

### ISSUE THREAD

**Issue by tocsoft** [SixLabors/ImageSharp.Web#8]
We want to refactor the code so that we produce 3 packages.

1. **`SixLabors.ImageSharp.Web.Core`** - the core processing engine with its own abstractions around the http request to normalise request formats between core and systemweb (should be netstandard only)
2. **`SixLabors.ImageSharp.Web.AspNetCore`** - the aspnet core middleware piece, this includes the extension methods for configuring in an aspnetcore style + the a middleware dispacher that calls out to the `SixLabors.ImageSharp.Web.Core` after translating the request context into an `ImageRequestContext` and handling the response.
3. **`SixLabors.ImageSharp.Web.SystemWeb`** - the full framework `HttpModule` based implementation like other version responsible for translating between `HttpContext` and `ImageRequestContext` and calling the processing pipeline.

_This will allow us to reach older installations targeting full framework and also would provide a transition option for those user using the older ImageProcessor.Web library allowing ImageProcessor.Web and ImageProcessor to be able to be mothballed as users would now have a transition option._

**Comment by JimBobSquarePants** [comment]
I think we should refactor but I'm not keen on introducing another core package in order to support what is essentially a legacy package. I don't want to rename our offering either.

I would much prefer to have a single solution with two projects sharing core code.

- **ImageSharp.Web** - Our main offering 
- **ImageSharp.Web.Compat** - Which offers legacy support for System.Web

That way we're not diluting our brand and are making it very clear that the System.Web support is for transitional purposes only.

**Comment by tocsoft** [comment]
So your suggesting using a code sharing (probably using a shared project) and compiling in the common code into each package? that would work to drop the `Core` package. I'm not sure its the right thing to do.. if we drop the `Core` package that would make it very difficult for third party devs to add plugins to the processing pipeline that could target all the frameworks.

I gut says we need to drop the straight IS.Web or at least make that the Core and produce webframework specific packages... AspNetCore, SystemWeb, NancyFx etc 

I don't feel the SystemWeb stuff should be targeted as "only be for transitional purposes"... its full framework and it isn't not actually going away anytime soon, its just not the shiny new framework, but you'll find that most devs (the silent corporate ones that don't jump onto the next bandwagon) will be using SystemWeb for a long time to come because it works and microsoft is still supporting it for a long while yet.

So in conclusion I feel it should be IS.Web.Core (or just IS.Web, but I think core is better here), then a series of IS.Web.{FrameWorkName} (AspNetCore, AspNet, NancyFx...) packages for each target web framework

**Comment by JimBobSquarePants** [comment]
Haha! I bloody hate having these conversations with you cos you're always right 😝 

Architecturally you're absolutely correct. We could add interfaces, helpers, and processors to the core library and add the implementations to the individual versions. 

Let's stick with the first two only though to start with and not overstretch ourselves. I want to keep processors to the core one's we've written so far also. Maybe we can get people to actually pay for others.

**Comment by JimBobSquarePants** [comment]
We'll have to be careful though with our interfaces.

I don't think we can actually reuse the resolvers/cache interfaces over the different platforms as for maximum flexibility I've used the `HttpContext` as a parameter. In some places I could replace with strings but things like the `IUriParser` would need to be capable of reading headers etc for maximum flexibility.

**Comment by JimBobSquarePants** [comment]
I'm going to close this as out of scope. If we get enough demand, maybe we'll add the functionality as a separate library.
